materialize-postgres/driver.go

```diff
@@ -83,22 +101,47 @@ func newPostgresDriver() pm.DriverServer {
 		ResourceSpecType: new(tableConfig),
 		NewResource:      func(sqlDriver.Endpoint) sqlDriver.Resource { return new(tableConfig) },
 		NewEndpoint: func(ctx context.Context, raw json.RawMessage) (sqlDriver.Endpoint, error) {
-			var parsed = new(config)
-			if err := pf.UnmarshalStrict(raw, parsed); err != nil {
+			var cfg = new(config)
+			if err := pf.UnmarshalStrict(raw, cfg); err != nil {
 				return nil, fmt.Errorf("parsing Postgresql configuration: %w", err)
 			}
 
 			log.WithFields(log.Fields{
-				"database": parsed.Database,
-				"address":  parsed.Address,
-				"user":     parsed.User,
+				"database": cfg.Database,
+				"address":  cfg.Address,
+				"user":     cfg.User,
 			}).Info("opening database")
 
-			db, err := sql.Open("pgx", parsed.ToURI())
+			// If SSH Endpoint is configured, then try to start a tunnel before establishing connections
+			if cfg.NetworkTunnel != nil && cfg.NetworkTunnel.SshForwarding != nil && cfg.NetworkTunnel.SshForwarding.SshEndpoint != "" {
+				host, port, err := net.SplitHostPort(cfg.Address)
+				if err != nil {
+					return nil, fmt.Errorf("splitting address to host and port: %w", err)
+				}
+
+				var sshConfig = &networkTunnel.SshConfig{
+					SshEndpoint: cfg.NetworkTunnel.SshForwarding.SshEndpoint,
+					PrivateKey:  []byte(cfg.NetworkTunnel.SshForwarding.PrivateKey),
+					ForwardHost: host,
+					ForwardPort: port,
+					LocalPort:   "5432",
+				}
+				var tunnel = sshConfig.CreateTunnel()
+
+				// FIXME/question: do we need to shut down the tunnel manually if it is a child process?
```

Comment:
I believe I made a similar comment on the `postgres-rc` PR here: I can't think of any reason why we would specifically need to call `tunnel.Stop` since if the connector shuts down the child process/container should also stop. I would hope that this wouldn't result in any situations where the container hangs on stopping.